packages/labs/router/src/routes.ts

```diff
@@ -163,14 +167,24 @@ export class Routes implements ReactiveController {
       // Simulate a tail group with the whole pathname
       this._currentParams = {0: tailGroup};
     } else {
-      const route = (this._currentRoute = this._getRoute(pathname));
+      const route = this._getRoute(pathname);
       if (route === undefined) {
         throw new Error(`No route found for ${pathname}`);
       }
       const pattern = getPattern(route);
-      const r = pattern.exec({pathname});
-      this._currentParams = r?.pathname.groups ?? {};
-      tailGroup = getTailGroup(this._currentParams);
+      const result = pattern.exec({pathname});
+      const params = result?.pathname.groups ?? {};
+      tailGroup = getTailGroup(params);
+      if (typeof route.enter === 'function') {
+        const success = await route.enter(params);
```

Comment:
This is a pit of failure for UX

Easy to block a route transition on a process that may be slow (and worse: fast in dev and slow in prod). In that case, as I read this, the user clicks a link and literally nothing happens

If we go with this design, we'll want to expose hooks so that apps can display a sign of life during the navigation (even if it's just an indeterminate progress bar along the top of the window)

Reply:
There's an intersection with Task here in that technically we have the same four async states: initial, pending, success, failure. I'd say for a router you really want to only expose pending and success to users (404 should be success for the 404 route).

This is one of the many things we need to collect use-cases for, probably as a user-story/design-requirement for the UX a developer is trying to create. Some cases will want to not do anything until the new route is ready to render. Native links are mostly like this (though we'll want to integrate with the [pending task protocol](https://github.com/webcomponents-cg/community-protocols/blob/main/proposals/pending-task.md) so a page could build a pending state affordance). Some cases will want to transition to a pending/loading UI, then transition to the rendered route.

Reply:
(nit, typo in the pending task protocol doc: https://github.com/webcomponents-cg/community-protocols/pull/28)

Reply:
Yeah, I think this is largely a thing to solve in defaults, affordances, and documentation

Ooh, looking at app history, we could potentially make use of the navigate event's transitionWhile method to use the browser's ordinary affordances for indicating that it's mid-navigation.

If that's the case though, I think that (at some point, perhaps not when app history initially launches) the navigation will probably also be cancellable by the user